core/src/main/scala/org/broadinstitute/dsde/rawls/snapshot/SnapshotService.scala

```diff
@@ -41,7 +46,29 @@ class SnapshotService(protected val userInfo: UserInfo, val dataSource: SlickDat
       val dataRepoReference = new DataRepoSnapshot().instanceName(terraDataRepoInstanceName).snapshot(snapshot.snapshotId)
       val ref = workspaceManagerDAO.createDataReference(workspaceContext.workspaceIdAsUUID, snapshot.name, snapshot.description, ReferenceTypeEnum.DATA_REPO_SNAPSHOT, dataRepoReference, CloningInstructionsEnum.NOTHING, userInfo.accessToken)
 
-      Future.successful(ref)
+      val datasetName = "deltalayer_" + ref.getReferenceId.toString.replace('-', '_')
+
+      val defaultIamRoles = Map(Acl.Role.OWNER -> Seq((clientEmail, Acl.Entity.Type.USER)), Acl.Role.WRITER -> Seq((deltaLayerStreamerEmail, Acl.Entity.Type.USER)))
+
+      val accessPolicies = Seq(SamWorkspacePolicyNames.owner, SamWorkspacePolicyNames.writer, SamWorkspacePolicyNames.reader)
+
+      val datasetLabels = Map("workspace_id" -> workspaceContext.workspaceId, "snapshot_id" -> snapshot.snapshotId)
+
+      // create BQ dataset, get workspace policies from Sam, and add those Sam policies to the dataset IAM
+      val IOResult = for {
+        samPolicies <- IO.fromFuture(IO(samDAO.listPoliciesForResource(SamResourceTypeNames.workspace, workspaceContext.workspaceId, userInfo)))
+        projectOwnerPolicy = samPolicies.filter(_.policyName == SamWorkspacePolicyNames.projectOwner).head.policy.memberEmails
+        filteredSamPolicies = samPolicies.filter(samPolicy => accessPolicies.contains(samPolicy.policyName)).map(_.email) ++ projectOwnerPolicy
+        samAclBindings = Acl.Role.READER -> filteredSamPolicies.map{ filteredSamPolicyEmail =>(filteredSamPolicyEmail, Acl.Entity.Type.GROUP) }.toSeq
+        aclBindings = defaultIamRoles + samAclBindings
+        bqService = bqServiceFactory.getServiceFromCredentialPath(pathToCredentialJson, GoogleProject(workspaceName.namespace))
+        _ <- bqService.use(_.createDataset(datasetName, datasetLabels, aclBindings))
+        petToken <- IO.fromFuture(IO(samDAO.getPetServiceAccountToken(GoogleProjectId(workspaceName.namespace), SamDAO.defaultScopes + SamDAO.bigQueryReadOnlyScope, userInfo)))
+      } yield { petToken }
+      IOResult.unsafeToFuture().map { petToken =>
+        workspaceManagerDAO.createBigQueryDataset(workspaceContext.workspaceIdAsUUID, new DataReferenceRequestMetadata().name(datasetName).cloningInstructions(CloningInstructionsEnum.NOTHING), new GoogleBigQueryDatasetUid().projectId(workspaceContext.namespace).datasetId(datasetName), OAuth2BearerToken(petToken))
+        ref
+      }
```

Comment:
This method is still vulnerable to orphaning resources. As I read it, this method:
1. Creates the snapshot reference
2. Creates BQ dataset
3. Creates dataset reference

(with other details elided, like getting policies from Sam and getting a pet token)

If any step fails, the previous steps will be orphaned. That is:
* if creating the BQ dataset fails, the user will have a snapshot reference without a Delta Layer companion
* if creating the dataset reference fails, the user will have an orphaned dataset in their workspace with no reference to it

The two paths I can see to solving this are to adopt Stairway, or to wrap this entire method in a try/catch or Future.recover, and in the failure case attempt to delete all three resources (snapshot reference, BQ dataset, dataset reference), ignoring any 404s from the delete attempts.

This cleanup code may drive changes to the accompanying tests, of course.